runtime/runtime/src/lib.rs

```diff
@@ -1709,47 +1689,124 @@ impl Runtime {
+        Ok(processed_delayed_receipts)
+    }
+
+    fn process_incoming_receipts<'a>(
+        &self,
+        mut processing_state: &mut ApplyProcessingReceiptState<'a>,
+        receipt_sink: &mut ReceiptSink,
+        compute_limit: u64,
+        proof_size_limit: Option<usize>,
+        validator_proposals: &mut Vec<ValidatorStake>,
+    ) -> Result<(), RuntimeError> {
         let incoming_processing_start = std::time::Instant::now();
-        if let Some(prefetcher) = &mut prefetcher {
+        let protocol_version = processing_state.protocol_version;
+        if let Some(prefetcher) = &mut processing_state.prefetcher {
             // Prefetcher is allowed to fail
-            _ = prefetcher.prefetch_receipts_data(&incoming_receipts);
+            _ = prefetcher.prefetch_receipts_data(&processing_state.incoming_receipts);
         }
-        for receipt in incoming_receipts.iter() {
+        for receipt in processing_state.incoming_receipts.iter() {
             // Validating new incoming no matter whether we have available gas or not. We don't
             // want to store invalid receipts in state as delayed.
             validate_receipt(
-                &apply_state.config.wasm_config.limit_config,
+                &processing_state.apply_state.config.wasm_config.limit_config,
                 receipt,
                 protocol_version,
             )
             .map_err(RuntimeError::ReceiptValidationError)?;
-            if total.compute >= compute_limit
+            if processing_state.total.compute >= compute_limit
                 || proof_size_limit.is_some_and(|limit| {
-                    state_update.trie.recorded_storage_size_upper_bound() > limit
+                    processing_state.state_update.trie.recorded_storage_size_upper_bound() > limit
                 })
             {
-                delayed_receipts.push(&mut state_update, receipt, &apply_state.config)?;
+                processing_state.delayed_receipts.push(
+                    &mut processing_state.state_update,
+                    receipt,
+                    &processing_state.apply_state.config,
+                )?;
             } else {
-                process_receipt_with_metrics(&receipt, state_update, total)?;
+                self.process_receipt_with_metrics(
+                    &receipt,
+                    &mut processing_state,
+                    receipt_sink,
+                    validator_proposals,
+                )?;
             }
         }
-        metrics.incoming_receipts_done(
-            incoming_receipts.len() as u64,
+        processing_state.metrics.incoming_receipts_done(
+            processing_state.incoming_receipts.len() as u64,
             incoming_processing_start.elapsed(),
-            total.gas,
-            total.compute,
+            processing_state.total.gas,
+            processing_state.total.compute,
         );
+        Ok(())
+    }
+
+    /// Processes all receipts (local, delayed and incoming).
```

Comment:
Technically there are also postponed receipts but I guess those are processed as part of other receipts processing.